edk2toollib/os/uefivariablesupport.py

```diff
@@ -0,0 +1,331 @@
+
+
+EFI_VAR_MAX_BUFFER_SIZE = 1024 * 1024
+
+
+class UefiVariable(object):
+    """Class to interact with Uefi Variables under Windows or Linux.
+
+    Methods:
+    -------
+    GetUefiVar - Get a single Uefi Variable's data
+    GetUefiAllVarNames - Get all Uefi variables
+    SetUefiVar - Set (or delete) a single Uefi variable.
+    """
+    ERROR_ENVVAR_NOT_FOUND = 0xcb
+
+    def __init__(self) -> None:
+        """Initialize Class."""
+        if os.name == 'nt':
+            # enable required SeSystemEnvironmentPrivilege privilege
+            privilege = win32security.LookupPrivilegeValue(
+                None, "SeSystemEnvironmentPrivilege"
+            )
+            token = win32security.OpenProcessToken(
+                win32process.GetCurrentProcess(),
+                win32security.TOKEN_READ | win32security.TOKEN_ADJUST_PRIVILEGES,
+            )
+            win32security.AdjustTokenPrivileges(
+                token, False, [(privilege, win32security.SE_PRIVILEGE_ENABLED)]
+            )
+            win32api.CloseHandle(token)
+
+            # import firmware variable API
+            try:
+                self._GetFirmwareEnvironmentVariable = (
+                    windll.kernel32.GetFirmwareEnvironmentVariableW
+                )
+                self._GetFirmwareEnvironmentVariable.restype = c_int
+                self._GetFirmwareEnvironmentVariable.argtypes = [
+                    c_wchar_p,
+                    c_wchar_p,
+                    c_void_p,
+                    c_int,
+                ]
+                self._EnumerateFirmwareEnvironmentVariable = (
+                    windll.ntdll.NtEnumerateSystemEnvironmentValuesEx
+                )
+                self._EnumerateFirmwareEnvironmentVariable.restype = c_int
+                self._EnumerateFirmwareEnvironmentVariable.argtypes = [
+                    c_int,
+                    c_void_p,
+                    c_void_p
+                ]
+                self._SetFirmwareEnvironmentVariable = (
+                    windll.kernel32.SetFirmwareEnvironmentVariableW
+                )
+                self._SetFirmwareEnvironmentVariable.restype = c_int
+                self._SetFirmwareEnvironmentVariable.argtypes = [
+                    c_wchar_p,
+                    c_wchar_p,
+                    c_void_p,
+                    c_int,
+                ]
+                self._SetFirmwareEnvironmentVariableEx = (
+                    windll.kernel32.SetFirmwareEnvironmentVariableExW
+                )
+                self._SetFirmwareEnvironmentVariableEx.restype = c_int
+                self._SetFirmwareEnvironmentVariableEx.argtypes = [
+                    c_wchar_p,
+                    c_wchar_p,
+                    c_void_p,
+                    c_int,
+                    c_int,
+                ]
+            except Exception:
+                logging.warn(
+                    "G[S]etFirmwareEnvironmentVariableW function doesn't seem to exist"
```

Comment:
Was this `G[S]et` on purpose or a typo?

Reply:
I think it was intentional.

The __init__ function is trying to collect the following functions:
GetFirmwareEnvironmentVariableW
NtEnumerateSystemEnvironmentValuesEx
SetFirmwareEnvironmentVariableW
SetFirmwareEnvironmentVariableExW

Its the same error message for any exception thrown. I agree it's confusing, so I will clarify it to something more generic like "Collecting windll Variable functions encountered an error."

Reply:
Fixed in latest.